8359963: compiler/c2/aarch64/TestStaticCallStub.java fails with for code cache > 250MB the static call stub is expected to be implemented using far branch

mikabl-arm opened the pull request:
The test assumed that hsdis is always available which is not the case. Make the test accept and scan either real or pseudo disassembly.

---------
- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8359963](https://bugs.openjdk.org/browse/JDK-8359963): compiler/c2/aarch64/TestStaticCallStub.java fails with for code cache &gt; 250MB the static call stub is expected to be implemented using far branch (**Bug** - P4)


### Contributors
 * Evgeny Astigeevich `<eastigeevich@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26047/head:pull/26047` \
`$ git checkout pull/26047`

Update a local copy of the PR: \
`$ git checkout pull/26047` \
`$ git pull https://git.openjdk.org/jdk.git pull/26047/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26047`

View PR using the GUI difftool: \
`$ git pr show -t 26047`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26047.diff">https://git.openjdk.org/jdk/pull/26047.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26047#issuecomment-3019630508)
</details>

mikabl-arm committed:
8359963: compiler/c2/aarch64/TestStaticCallStub.java fails with for code cache > 250MB the static call stub is expected to be implemented using far branch

The test assumed that hsdis is always available which is not the case.
Make the test accept and scan either real or pseudo disassembly.

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -97,19 +304,19 @@ static void verifyFarStaticCall(ListIterator<String> itr) {
 
         ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(procArgs);
         OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        System.out.println(output.getOutput());
```
I think it is worth to have `output.shouldHaveExitValue(0);`.

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +48,254 @@
-            }
+        while (itr.hasNext() && extracted.size() < n) {
+            int left = n - extracted.size();
+            extractOpcodeOrBytecodes(itr.next()).stream().limit(left).forEach(extracted::add);
```
You can detect whether you have hex codes or disassembly.
See https://github.com/openjdk/jdk/blob/a86dd56de34f730b42593236f17118ef5ce4985a/test/hotspot/jtreg/compiler/onSpinWait/TestOnSpinWaitAArch64.java#L122

In TestOnSpinWaitAArch64 you can see how the checking code is organized not to depend on the instruction representation.

IMO you don't need the Instruction class hierarchy.
You need `nearStaticCallOpcodeSeq` and `farStaticCallOpcodeSeq` to be filled either with opcodes or hex codes. Of course their names need to be changed to something like `nearStaticCallInstSeq`. You will need to change `extractOpcodesN` and `extractOpcode` to `extractInstructionsN` and `extractInstruction`.

mikabl-arm committed:
Merge branch 'master' at d46298d6 into 8359963

Change-Id: I8a405439e3b64b4cd3af80458b406b1f3dcefe42

mikabl-arm commented:
/open

mikabl-arm committed:
Address review comments

- Add output.shouldHaveExitValue(0)
- Simplify the implementation by giving up the Instruction class hierarchy

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +47,136 @@
+        new Instruction("movk", 0xFF800000, 0xF2800000), // MOVK
+        new Instruction("movk", 0xFF800000, 0xF2800000), // MOVK
+        new Instruction("br",   0xFFFFFC1F, 0xD61F0000)  // BR
+    };
```
nit:
```
private static final Instruction ISB  = new Instruction("isb",  0xFFFFF0FF, 0xD50330DF);
private static final Instruction MOVK = new Instruction("movk", 0xFF800000, 0xF2800000);
private static final Instruction MOVZ = new Instruction("mov",  0xFF800000, 0xD2800000);
private static final Instruction B = new Instruction("b",    0xFC000000, 0x14000000);
private static final Instruction BR = new Instruction("br",   0xFFFFFC1F, 0xD61F0000);

private static final Instruction[] nearStaticCallInsts = {ISB, MOVZ, MOVK, MOVK, B};
private static final Instruction[] farStaticCallInsts = {ISB, MOVZ, MOVK, MOVK, MOVZ, MOVK, MOVK, BR};
```

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +47,136 @@
+            return words[0].trim();
         }
 
         return "";
```
As `isBlank()` is false here, `if (words.length > 0)` is always true. Also `trim()` is redundant. So you need just one `return line.split("\\s+")[0];`.

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +47,136 @@
+        int semicolonIndex = line.indexOf(';');
+        if (semicolonIndex != -1) {
+            line = line.substring(0, semicolonIndex).trim();
+        }
```
The same code is in `extractMnemonic`. It's worth to extract into a method.

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
This can be rewritten without `disassembled`:
```
interface StaticCallMatcher {
    public boolean match(ListIterator<String> iter, Instruction[] insts);
}

static class StaticCallMnemonicMatcher implements StaticCallMatcher {
     public boolean match(ListIterator<String> iter, Instruction[] insts) {
           return extractMnemonicsN(iter, insts.length).equals(Arrays.stream(insts).map(Instruction::mnemonic).toList());
     }
}

static class StaticCallOpcodeMatcher implements StaticCallMatcher {
     public boolean match(ListIterator<String> iter, Instruction[] insts) {
            List<Integer> extracted = extractOpcodesN(iter, insts.length);
            return opcodesMatch(extracted, insts);
     }
}

static abstract class StaticCallVerifier {
    private StaticCallMatcher matcher;
    Instruction[] insts;

    protected StaticCallVerifier(StaticCallMatcher matcher, Instruction[] insts) {
        this.matcher = matcher;
        this.insts = insts;
    }

    protected abstract void error();

    public void verify(ListIterator<String> iter) {
          if (!matcher.match(iter, instrs) {
               error();
          }
    }
}

static class NearStaticCallVerifier extends StaticCallVerifier {
      public NearStaticCallVerifier(StaticCallMatcher matcher) {
             super(matcher, nearStaticCallInsts);
      }
      protected void error() {
            throw new RuntimeException(
                    "for code cache < 250MB the static call stub is expected to be implemented using near branch");
      }  
}

static class FarStaticCallVerifier extends StaticCallVerifier {
      public FarStaticCallVerifier(StaticCallMatcher matcher) {
             super(matcher, farStaticCallInsts);
      }
      protected void error() {
            throw new RuntimeException(
                    "for code cache > 250MB the static call stub is expected to be implemented using far branch");
      }  
}
```

eastig replied:
Too many classes...

Maybe this would be simpler:
```
@FunctionalInterface
interface StaticCallMatcher {
    boolean matches(ListIterator<String> iter, Instruction[] insts);
}

static boolean matchMnemonics(ListIterator<String> iter, Instruction[] insts) {
    return extractMnemonicsN(iter, insts.length)
            .equals(Arrays.stream(insts).map(Instruction::mnemonic).toList());
}

static boolean matchOpcodes(ListIterator<String> iter, Instruction[] insts) {
    return opcodesMatch(extractOpcodesN(iter, insts.length), insts);
}

static void verifyStaticCall(ListIterator<String> iter, StaticCallMatcher matcher,
                             Instruction[] insts, String errorMessage) {
    if (!matcher.matches(iter, insts)) {
        throw new RuntimeException(errorMessage);
    }
}
```

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -97,35 +185,50 @@ static void verifyFarStaticCall(ListIterator<String> itr) {
                 }
-                return;
             }
+
```
```
StaticCallMatcher staticCallMatcher = (disassembled) ? new StaticCallMnemonicMatcher() : new StaticCallOpcodeMatcher();
StaticCallVerifier staticCallVerifier = (bigCodeCache) ? new FarStaticCallVerifier(staticCallMatcher) : new NearStaticCallVerifier(staticCallMatcher);
```

eastig replied:
Variant for the simpler version:
```
StaticCallMatcher matcher = disassembled ? TestStaticCallStub::matchMnemonics
                                         : TestStaticCallStub::matchOpcodes;
```

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -97,35 +185,50 @@ static void verifyFarStaticCall(ListIterator<String> itr) {
+            // 2. Look for the block comment
+            while (iter.hasNext()) {
+                String line = iter.next();
+                if (line.contains("{static_stub}")) {
+                    iter.previous();
+                    if (bigCodeCache) {
+                        verifyFarStaticCall(iter, disassembled);
+                    } else {
+                        verifyNearStaticCall(iter, disassembled);
+                    }
```
```
staticCallVerifier.verify(iter);
```

eastig replied:
Variant for the simpler version:
```
if (bigCodeCache) {
    verifyStaticCall(iter, matcher, farStaticCallInsts,
            "for code cache > 250MB the static call stub is expected to be implemented using far branch");
} else {
    verifyStaticCall(iter, matcher, nearStaticCallInsts,
            "for code cache < 250MB the static call stub is expected to be implemented using near branch");
}
```

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +47,136 @@
+        new Instruction("b",    0xFC000000, 0x14000000)  // B
+    };
+    private static final Instruction[] farStaticCallInsts = {
+        new Instruction("isb",  0xFFFFF0FF, 0xD50330DF), // ISB
```
Why do we need `isb` in this form instead of `new Instruction("isb",  0xFFFFFFFF, 0xD5033FDF)`?

eastig commented on test/hotspot/jtreg/compiler/c2/aarch64/TestStaticCallStub.java:
```diff
@@ -43,52 +47,136 @@
+            int value = (Integer.parseUnsignedInt(halfwords[0].trim(), 16) << 16)
+                    + Integer.parseUnsignedInt(halfwords[1].trim(), 16);
+            value = Integer.reverseBytes(value);
+            opcodes.add(value);
```
If a regression happens: a stub for a near call stub is created but we expect a far call stub, this code will cause a problem. `NumberFormatException` will be thrown instead of a correct runtime exception.

Example:
```
  0x0000000105edeb90: ;   {static_stub}
  0x0000000105edeb90: df3f 03d5 

  0x0000000105edeb94: ;   {metadata(nullptr)}
  0x0000000105edeb94: 0c00 80d2 | 0c00 a0f2 | 0c00 c0f2 | 0000 0014 
[Exception Handler]
```

We expect 8 instructions but 5 instructions has been generated. So `extractOpcodes` will be called for `[Exception Handler]` which will cause `Integer.parseUnsignedInt` to fail with NumberFormatException`.

eastig replied:
BTW, I think this is simpler:
```
int value = Integer.reverseBytes(Integer.parseUnsignedInt(word.replaceAll("\\s", ""), 16));
```

The fix of `NumberFormatException`:
```
String hexValue = word.replaceAll("\\s", "");
if (!hexValue.matches("\\p{XDigit}{8}")) {
     continue;
}
opcodes.add(Integer.reverseBytes(Integer.parseUnsignedInt(hexValue, 16)));
```

mikabl-arm replied:
> We expect 8 instructions but 5 instructions has been generated. So extractOpcodes will be called for [Exception Handler] which will cause Integer.parseUnsignedInt to fail with NumberFormatException`.

It won't, because `instructionsSubstring()` returns an empty string for the `[Exception Handler]` line.

I've verified this behavior by replacing `"256M"` in `procArgs` with `"200M"`. This makes the VM generate near static call stubs while the test still expects a far call stub.

(ignore line numbers)

```
[Stub Code]
  0x0000f11bb5103e10: ;   {trampoline_stub}
  0x0000f11bb5103e10: 4800 0058 | 0001 1fd6 | 403c 7cb4 | 1bf1 0000

  0x0000f11bb5103e20: ;   {trampoline_stub}
  0x0000f11bb5103e20: 4800 0058 | 0001 1fd6 | 403c 7cb4 | 1bf1 0000

  0x0000f11bb5103e30: ;   {static_stub}
  0x0000f11bb5103e30: df3f 03d5

  0x0000f11bb5103e34: ;   {metadata(nullptr)}
  0x0000f11bb5103e34: 0c00 80d2 | 0c00 a0f2 | 0c00 c0f2 | 0000 0014

  0x0000f11bb5103e44: ;   {runtime_call DeoptimizationBlob}
  0x0000f11bb5103e44: 9f04 db97
[Deopt Handler Entry Point]
  0x0000f11bb5103e48: ffff ff17 | 0000 0000
[/MachCode]
Ok

STDERR:
java.lang.RuntimeException: for code cache > 250MB the static call stub is expected to be implemented using far branch
        at compiler.c2.aarch64.TestStaticCallStub.verifyStaticCall(TestStaticCallStub.java:152)
        at compiler.c2.aarch64.TestStaticCallStub.runVM(TestStaticCallStub.java:194)
        at compiler.c2.aarch64.TestStaticCallStub.main(TestStaticCallStub.java:213)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:583)
        at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:335)
        at java.base/java.lang.Thread.run(Thread.java:1527)
```

mikabl-arm replied:
I realized that I haven't submitted `instructionSubstring()` I was referring to yet. It would be a method common string manipulation logic is extracted to as you suggested above.

```
        int colonIndex = line.indexOf(':');
        if (colonIndex != -1) {
            line = line.substring(colonIndex + 1).trim();
        }

        int semicolonIndex = line.indexOf(';');
        if (semicolonIndex != -1) {
            line = line.substring(0, semicolonIndex).trim();
        }
```

eastig replied:
> It won't, because instructionsSubstring() returns an empty string for the [Exception Handler] line.

How can `instructionsSubstring("[Exception Handler]")` return an empty string? The input string is modified only if it contains either ':' or ';'. "[Exception Handler]" contains neither of them.

mikabl-arm committed:
integrate changes contributed by @eastig

mikabl-arm committed:
extract common string manipulation logic into instructionsSubstring()

mikabl-arm committed:
always expect CRm = 0b1111 in ISBs

mikabl-arm commented:
/contributor add @eastig

mikabl-arm commented:
/touch
